### PR TITLE
WIP: Chef-16 resource cleanup + unified_mode

### PR DESCRIPTION
**Pull request by lamont-granquist**
* apt_package: unified_mode
* apt_preference: unified_mode + converted to custom resource
* apt_repository: unified_mode + converted to custom resource
* apt_update: unified_mode + converted to custom resource
* archive_file: unified_mode
* bff_package: unified_mode
* build_essential: unified_mode
* cab_package: unified_mode
* chef_gem: unified_mode
* chef_handler: unified_mode
* chef_sleep: unified_mode
* chocolatey_config: unified_mode
* chocolatey_feature: unified_mode
* chocolatey_package: unified_mode
* chocolatey_source: unified_mode
* cookbook_file: unified_mode
* cron: unified_mode
* cron_access: unified_mode
* cron_d: unified_mode
* dmg_package: unified_mode
* dpkg_package: unified_mode
* dsc_script: unified_mode
* freebsd_package: unified_mode
* gem_package: unified_mode

### CODE REVIEW

File: lib/chef/resource/apt_preference.rb [hunk 1]
```diff
@@ -47,6 +49,72 @@ class AptPreference < Chef::Resource
...
+        )
+
+        directory APT_PREFERENCE_DIR do
+          mode "0755"
+          action :create
+        end
+
+        sanitized_prefname = safe_name(new_resource.package_name)
+
+        # cleanup any existing pref files w/o the sanitized name (created by old apt cookbook)
```
Comment (tas50) [inline lib/chef/resource/apt_preference.rb]: We should probably nuke this in Chef 17. Might want to add todo comment there.

File: lib/chef/resource/apt_preference.rb [hunk 2]
```diff
@@ -47,6 +49,72 @@ class AptPreference < Chef::Resource
...
+
+        file "#{APT_PREFERENCE_DIR}/#{sanitized_prefname}.pref" do
+          mode "0644"
+          content preference
+          action :create
+        end
+      end
+
+      action :remove do
+        return unless debian?
```
Comment (tas50) [inline lib/chef/resource/apt_preference.rb]: Is this to preserve the provides bit w/o the complexity?

File: lib/chef/resource/build_essential.rb [hunk 3]
```diff
@@ -56,11 +58,11 @@ class BuildEssential < Chef::Resource
...
         when debian?
           package %w{ autoconf binutils-doc bison build-essential flex gettext ncurses-dev }
-         when fedora_derived?
-           package %w{ autoconf bison flex gcc gcc-c++ gettext kernel-devel make m4 ncurses-devel patch }
+        when fedora_derived?
+          package %w{ autoconf bison flex gcc gcc-c++ gettext kernel-devel make m4 ncurses-devel patch }
 
           # Ensure GCC 4 is available on older pre-6 EL
-           package %w{ gcc44 gcc44-c++ } if platform_family?("rhel") && node["platform_version"].to_i < 6
+          package %w{ gcc44 gcc44-c++ } if platform_family?("rhel") && node["platform_version"].to_i < 6
```
Comment (tas50) [inline lib/chef/resource/build_essential.rb]: we can probably just remove this RHEL 5 bit while we're here

File: lib/chef/resource/apt_preference.rb (hunk 2 above)
Comment (lamont-granquist) [reply lib/chef/resource/apt_preference.rb]: yeah, the old Noop provider method used the provider resolver in a way that made it impossible to write this as a custom resource.  in order to convert it that had to be abandoned.  the Noop provider trick wasn't one of my best ideas.

File: lib/chef/resource/apt_preference.rb (hunk 1 above)
Comment (lamont-granquist) [reply lib/chef/resource/apt_preference.rb]: nope, out of scope

File: lib/chef/resource/build_essential.rb (hunk 3 above)
Comment (lamont-granquist) [reply lib/chef/resource/build_essential.rb]: nope, out of scope

**Pull request merged**